pyro/contrib/timeseries/lgssm.py

```diff
@@ -0,0 +1,118 @@
+                                self.obs_matrix, self._get_obs_dist())
+
+    def log_prob(self, targets):
+        """
+        :param torch.Tensor targets: A 2-dimensional tensor of real-valued targets
+            of shape `(T, obs_dim)`, where `T` is the length of the time series and `obs_dim`
+            is the dimension of the real-valued `targets` at each time step
+        :returns torch.Tensor: A (scalar) log probability.
+        """
+        assert targets.dim() == 2 and targets.size(-1) == self.obs_dim
+        return self._get_dist().log_prob(targets)
+
+    def _filter(self, targets):
+        """
+        Return the filtering state for the associated state space model.
+        """
+        assert targets.dim() == 2 and targets.size(-1) == self.obs_dim
+        return self._get_dist().filter(targets)
+
+    def _forecast(self, N_timesteps, filtering_state, include_observation_noise=True):
+        """
+        Internal helper for forecasting.
+        """
+        N_trans_matrix = repeated_matmul(self.trans_matrix, N_timesteps)
+        N_trans_obs = torch.matmul(N_trans_matrix, self.obs_matrix)
+        predicted_mean = torch.matmul(filtering_state.loc.unsqueeze(-2), N_trans_obs).squeeze(-2)
+
+        # first compute the contribution from filtering_state.covariance_matrix
+        predicted_covar1 = torch.matmul(N_trans_obs.transpose(-1, -2),
+                                        torch.matmul(filtering_state.covariance_matrix,
+                                        N_trans_obs))  # N O O
+
+        # next compute the contribution from process noise that is injected at each timestep.
+        # (we need to do a cumulative sum to integrate across time)
+        process_covar = self._get_trans_dist().covariance_matrix
+        N_trans_obs_shift = torch.cat([self.obs_matrix.unsqueeze(0), N_trans_obs[0:-1]])
+        predicted_covar2 = torch.matmul(N_trans_obs_shift.transpose(-1, -2),
+                                        torch.matmul(process_covar, N_trans_obs_shift))  # N O O
+
+        predicted_covar = predicted_covar1 + torch.cumsum(predicted_covar2, dim=0)
+
+        if include_observation_noise:
+            eye = torch.eye(self.obs_dim, device=self.obs_matrix.device, dtype=self.obs_matrix.dtype)
+            predicted_covar = predicted_covar + self._get_obs_noise_scale().pow(2.0) * eye
+
+        return predicted_mean, predicted_covar
+
+    def forecast(self, targets, N_timesteps):
+        """
+        :param torch.Tensor targets: A 2-dimensional tensor of real-valued targets
+            of shape `(T, obs_dim)`, where `T` is the length of the time series and `obs_dim`
+            is the dimension of the real-valued targets at each time step. These
+            represent the training data that are conditioned on for the purpose of making
+            forecasts.
+        :param int N_timesteps: The number of timesteps to forecast into the future from
```

Comment:
changed the API here as compared to the (continuous-time) GP models. what is the best way to do this? just expand the docstring in the base class?